Implement RandomNumberGenerator.GetBytes

vcsjones opened the pull request:
Closes #42769

vcsjones committed:
Implement RandomNumberGenerator.GetBytes(int).

vcsjones committed:
Use RandomNumberGenerator.GetBytes(int) where possible.

stephentoub commented on src/libraries/System.Security.Cryptography.Algorithms/ref/System.Security.Cryptography.Algorithms.cs:
```diff
@@ -544,6 +544,7 @@ public abstract partial class RandomNumberGenerator : System.IDisposable
         public static void Fill(System.Span<byte> data) { }
         public abstract void GetBytes(byte[] data);
         public virtual void GetBytes(byte[] data, int offset, int count) { }
+        public static byte[] GetBytes(int byteCount) { throw null; }
```
Minor, but just want to check: @GrabYourPitchforks, this argument name could just be `count` to match the `count` on the other GetBytes overload... did we consider that?  I'm ok either way.

GrabYourPitchforks replied:
TBH I don't have a strong stance on what the arg name should be. If _count_ flows better then so be it. :)

vcsjones replied:
I am hearing I should change it then?

bartonjs replied:
The only other `GetBytes(int)` that I'm seeing (other than BitConverter.GetBytes, which has a very different meaning) calls the parameter `cb`, which is not to be emulated.

If someone used named syntax then `byte[] random = RandomNumberGenerator.GetBytes(count: 32);` looks more fluid than `byte[] random = RandomNumberGenerator.GetBytes(byteCount: 32);`... so I guess I support simplifying the name to `count`.

stephentoub commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs:
```diff
@@ -146,6 +146,29 @@ public static int GetInt32(int toExclusive)
+                throw new ArgumentOutOfRangeException(nameof(byteCount), SR.ArgumentOutOfRange_NeedNonNegNum);
+
+            if (byteCount == 0)
+                return Array.Empty<byte>();
```
Nit: consider swapping the order of these ifs.  If 0 is passed in, then we won't need to do the < 0 check.  If anything else is passed in, we need to do both checks anyway.

vcsjones commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs:
```diff
@@ -146,6 +146,29 @@ public static int GetInt32(int toExclusive)
+            if (byteCount == 0)
+                return Array.Empty<byte>();
+
+            byte[] ret = GC.AllocateUninitializedArray<byte>(byteCount);
```
That's a good point. I would expect that most usages would probably fall in that `< 2048` path and just hit `new T[]`, at least in places where cryptographic values are used. I guess if I wanted a huge array I could always allocate it myself and use `Fill`.

I suppose then to ask, is there any harm if AUA is used since it falls back, or does it still have some non-consequential overhead for small arrays? Is the threshold of "use new T[]" an implementation detail? (Could it get faster for small arrays at some point?)

vcsjones commented on src/libraries/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs:
```diff
@@ -229,6 +229,35 @@ public static void GetBytes_InvalidArgs_Base()
+        }
+
+        [Fact]
+        public static void GetBytes_Int_RandomDistrobution()
```
Nit for myself that I will fix pending an open question: small Distribution correctly.

vcsjones committed:
Code review feedback and various typos.

GrabYourPitchforks commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs:
```diff
@@ -146,6 +146,29 @@ public static int GetInt32(int toExclusive)
+        /// </exception>
+        public static byte[] GetBytes(int byteCount)
+        {
+            if (byteCount == 0)
```
Honestly I'd remove this if check. There's no point to optimizing the zero-length case because nobody will realistically ever pass 0 as a parameter.

vcsjones committed:
More code review feedback.

Remove Array.Empty check as it is not expected to be a common path worth
handling.
Change parameter name.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/ref/System.Security.Cryptography.Algorithms.cs:
```diff
@@ -544,6 +544,7 @@ public abstract partial class RandomNumberGenerator : System.IDisposable
         public static void Fill(System.Span<byte> data) { }
         public abstract void GetBytes(byte[] data);
         public virtual void GetBytes(byte[] data, int offset, int count) { }
+        public static byte[] GetBytes(int byteCount) { throw null; }
```
Looks like only the implementation got the parameter name update, missed it in the ref :)

vcsjones replied:
Huh you're right. I would have expected that to break the `ApiCompat` check but `dotnet build` doesn't complain. :shrug: .

bartonjs replied:
I believe ApiCompat only sees the type and modifiers from a signature, and they were both `public static byte[] GetBytes(System.Int32)`.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs:
```diff
@@ -146,6 +146,26 @@ public static int GetInt32(int toExclusive)
+        /// An array populated with cryptographically strong random values.
+        /// </returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// The <paramref name="count" /> parameter is less than zero.
```
Based on the ArgumentOutOfRangeException example at https://github.com/dotnet/dotnet-api-docs/wiki/Exceptions, I think this should be

```suggestion
        /// <paramref name="count" /> is less than zero.
```

vcsjones committed:
Fix parameter name in ref

vcsjones committed:
Feedback on documentation
